catalog/transformer/catalog-transformer-streaming-lib/src/main/java/org/codice/ddf/transformer/xml/streaming/lib/MetacardTypeRegister.java

```diff
@@ -0,0 +1,122 @@
+
+    private List<SaxEventHandlerFactory> saxEventHandlerFactories = new ArrayList<>();
+
+    private MetacardTypeImpl metacardType;
+
+    private String id = "DEFAULT_ID";
+
+    public synchronized void bind(SaxEventHandlerFactory saxEventHandlerFactory) {
+        saxEventHandlerFactories.add(saxEventHandlerFactory);
+        setMetacardType();
+    }
+
+    public synchronized void unbind(SaxEventHandlerFactory saxEventHandlerFactory) {
+        saxEventHandlerFactories.remove(saxEventHandlerFactory);
+        setMetacardType();
+    }
+
+    /**
+     * Defines a {@link MetacardTypeImpl} based on component Sax Event Handler Factories
+     * and what attributes they populate
+     *
+     * @return a DynamicMetacardType that describes the type of metacard that is created in this transformer
+     */
+    private synchronized void setMetacardType() {
+        Set<AttributeDescriptor> attributeDescriptors = new HashSet<>();
+
+        for (SaxEventHandlerFactory factory : saxEventHandlerFactories) {
+            attributeDescriptors.addAll(factory.getSupportedAttributeDescriptors());
+        }
+        attributeDescriptors.addAll(BasicTypes.BASIC_METACARD.getAttributeDescriptors());
+
+        MetacardTypeImpl dynamicMetacardType = new MetacardTypeImpl(id, attributeDescriptors);
+        registerMetacardType(dynamicMetacardType);
+        metacardType = dynamicMetacardType;
+    }
+
+    public MetacardType getMetacardType() {
+        if (metacardType == null) {
+            setMetacardType();
+        }
+
+        return metacardType;
+    }
+
+    private synchronized void registerMetacardType(MetacardType metacardType) {
+        unregisterMetacardType(metacardType);
+
+        Dictionary serviceProperties = new Hashtable();
+        serviceProperties.put("name", metacardType.getName());
+
+        ServiceRegistration serviceRegistration =
+                getContext().registerService(MetacardType.class.getName(),
+                        metacardType,
+                        serviceProperties);
+
+        this.metacardTypeServiceRegistrations.put(metacardType.getName(), serviceRegistration);
+    }
+
+    private synchronized void unregisterMetacardType(MetacardType metacardType) {
+        if (metacardType == null) {
+            return;
+        }
+
+        ServiceRegistration serviceRegistration =
+                metacardTypeServiceRegistrations.get(metacardType.getName());
+        if (serviceRegistration != null) {
+            serviceRegistration.unregister();
+            metacardTypeServiceRegistrations.remove(metacardType.getName());
+        }
+    }
+
+    public synchronized void setId(String id) {
+        this.id = id;
+    }
+
+    protected BundleContext getContext() {
+        Bundle cxfBundle = FrameworkUtil.getBundle(SortedServiceList.class);
+        if (cxfBundle != null) {
+            return cxfBundle.getBundleContext();
+        }
+        return null;
```

Comment:
You might want to consider throwing an exception instead of returning null - right now a `NullPointerException` would be thrown if we can't get the bundle context. And `MetacardTypeRegister.class` should be passed to `getBundle()`.